File: src/main/java/com/elpassion/problem/ProblemFactory.java
package com.elpassion.problem;

import java.util.UUID;

public class ProblemFactory {

    private final String name;
    private final UUID uuid;

    public String getName() {
        return name;
    }

    public UUID getUuid() {
        return uuid;
    }

    public ProblemFactory(String name) {
        this.name = name;
        this.uuid = UUID.randomUUID();
    }

    static ProblemFactory createProblemFactory(String name) {
        return new ProblemFactory(name);
    }

    public void createProblem() {
        throw new UnexpectedChild(name + " " + uuid);
    }
}

File: src/main/java/com/elpassion/problem/ProblemFactory.kt
package com.elpassion.problem

import java.util.UUID

class ProblemFactory(val name: String) {
    val uuid: UUID

    init {
        this.uuid = UUID.randomUUID()
    }

    fun createProblem() {
        throw UnexpectedChild(name + " " + uuid)
    }

    companion object {

        internal fun createProblemFactory(name: String): ProblemFactory {
            return ProblemFactory(name)
        }
    }
}
